### PR TITLE
large bin attack for glibc2.30

### PR DESCRIPTION
**Pull request by jwang-a**
The old large bin insertion got two additional checks in glibc2.30.
However, an alternative attack path still exist in latest versions (currently glibc2.30 ~ glibc2.32)
This specific path has existed since before glibc2.23, but I personally have never seen it discussed publicly.

### CODE REVIEW

**Comment by Kyle-Kyle** [comment]
I haven't looked into the technique yet. But by skimming through the technique, I have a question: is this `house-of-storm`? At least the relevant code snippet is the same as what `house-of-storm` attacks.
If they are the same, I'd like to rename it to `house-of-storm`.

**Comment by jwang-a** [comment]
**TL;DR**

- Is the relevant code snippet same as `house-of-storm`?
    No. It might seem similar at first glance because both snippets handle similar stuff for different cases.

- Is this attack in some sense similar to `house-of-storm`?
    Yes, since `house-of-storm` is an application of `large-bin-attack`.


**Details**

Correct me if I'm wrong, but I believe house of storm is just an application of large bin attack. It might better to just leave it as large bin attack, since there are many possible use case of this primitive aside from `house-of-storm`.

And just to clarify thing up, the value of this PoC is not a "new exploit vector", but rather to bypass current checks and revive `large-bin-attack` with an relatively unknown code path.

Generally speaking, there are three major cases when inserting large bin chunks :
1. victim chunk size < smallest chunk in bin
2. victim chunk size exist in bin
3. elsewise

Previous attacks all focus on the 3rd case (this includes `house-of-storm` and a lot of other challenges). However, starting from glibc2.30, new checks have been enforced and the 3rd case doesn't work anymore. The PoC presented here takes the 1st case instead, which currently has no security checks.

**Comment by Kyle-Kyle** [comment]
This is confirmed to be working on 2.31. Will be merged after the refactoring.
thanks!

**Comment by mdulin2** [comment]
I appreciate that this was added to glibc_2.30 folder. I think it is good to get the the versioning separate to patches to existing attacks.

**Comment by mdulin2** [comment]
Should a POC for the _House of Storm_ be added then? This looks like an existing technique, that works very well, that is not in the repo.

**Comment by Kyle-Kyle** [comment]
@mdulin2 we are doing it. checkout refactor branch :D
Yes. House of Storm should be added. I just don't have the bandwidth to do it.
